### ISSUE TITLE
[BUG] ParabolicSarIndicator throws NPE from getValue() method

### ISSUE THREAD

**Issue by zscauer** [ta4j/ta4j#973]
`getValue()` method of `ParabolicSarIndicator` works, till bar series don't remove second bar from series. To reproduce bug I have wrote this small test:

```Java
    @Test
    public void growingBarSeriesTest() {
        List<Bar> bars = new ArrayList<>();
        bars.add(new MockBar(4120.98, 4086.29, 4211.08, 4032.62, numFunction));

        MockBarSeries mockBarSeries = new MockBarSeries(bars);
        mockBarSeries.setMaximumBarCount(4);

        mockBarSeries.addBar(new MockBar(4069.13, 4016.00, 4119.62, 3911.79, numFunction));
        mockBarSeries.addBar(new MockBar(4016.00, 4040.00, 4104.82, 3400.00, numFunction));
        mockBarSeries.addBar(new MockBar(4040.00, 4114.01, 4265.80, 4013.89, numFunction));
        mockBarSeries.addBar(new MockBar(4147.00, 4316.01, 4371.68, 4085.01, numFunction));
        mockBarSeries.addBar(new MockBar(4316.01, 4280.68, 4453.91, 4247.48, numFunction));

        ParabolicSarIndicator sar = new ParabolicSarIndicator(mockBarSeries);

        assertThrows(NullPointerException.class, () -> sar.getValue(mockBarSeries.getEndIndex()));
        assertThrows(NullPointerException.class, () -> sar.getValue(mockBarSeries.getEndIndex() - 1));
        assertThrows(NullPointerException.class, () -> sar.getValue(mockBarSeries.getEndIndex() - 2));
        assertThrows(NullPointerException.class, () -> sar.getValue(mockBarSeries.getEndIndex() - 3));
    }
```

**Comment by team172011** [comment]
This is related to the caching logic and would be fixed by a endTime based cache like suggested in PR #907

Another fix would be to change the maps that are used for extreme point and trend detection from Map<Integer, > to Map<ZonedDateTime, > using the endTime of the bar instead of the index

**Comment by team172011** [comment]
Suggestion of a fix in #975

**Comment by team172011** [comment]
I updated #975 with the only solution I could imagine at the moment. This approach is recalculating all values from the start of the BarSeries to always have the maps filled with the correct values for the current index:

```java
    @Override
    protected Num calculate(int index) {
        lastExtreme.clear();
        lastAf.clear();
        isUpTrendMap.clear();

        // Caching of this indicator value calculation is essential for performance!
        //
        // clear the maps and recalculate the values for start to index
        // the internal calculations until the previous index will fill the
        // required maps for the acceleration factor, the trend direction and the
        // last extreme value
        for(int i = getBarSeries().getBeginIndex(); i < index; i++) {
             calculateInternal(i);
        }

        return calculateInternal(index);

    }

private Num calculateInternal(int index) {
        Num sar = NaN;
        boolean is_up_trend;

        if (index == getBarSeries().getBeginIndex()) {
            lastExtreme.put(index, getBarSeries().getBar(index).getClosePrice());
            lastAf.put(index, zero());
            isUpTrendMap.put(index, false);
            return sar; // no trend detection possible for the first value
        } else if (index == getBarSeries().getBeginIndex() + 1) {// start trend detection
            is_up_trend = getBarSeries().getBar(index - 1)
                    .getClosePrice()
                    .isLessThan(getBarSeries().getBar(index).getClosePrice());

            lastAf.put(index, accelerationStart);
            isUpTrendMap.put(index, is_up_trend);
            if (is_up_trend) { // up trend
                sar = new LowestValueIndicator(lowPriceIndicator, 2).getValue(index - 1); // put the lowest low value of
                // two
                lastExtreme.put(index, new HighestValueIndicator(highPriceIndicator, 2).getValue(index - 1));
            } else { // down trend
                sar = new HighestValueIndicator(highPriceIndicator, 2).getValue(index - 1); // put the highest high
                // value of
                lastExtreme.put(index, new LowestValueIndicator(lowPriceIndicator, 2).getValue(index - 1));
            }
            return sar;
        }

        Num priorSar = getValue(index - 1);

        is_up_trend = isUpTrendMap.get(index - 1);

        Num currentExtremePoint = lastExtreme.get(index - 1);
        Num cur_high = highPriceIndicator.getValue(index);
        Num cur_low = lowPriceIndicator.getValue(index);
        Num cur_af = lastAf.get(index - 1);
        sar = priorSar.plus(cur_af.multipliedBy((currentExtremePoint.minus(priorSar))));

        if (is_up_trend) { // if up trend
            if (cur_low.isLessThan(sar)) { // check if sar touches the low price
                sar = currentExtremePoint;

                lastAf.put(index, accelerationStart);
                lastExtreme.put(index, cur_low);
                is_up_trend = false;

            } else { // up trend is going on
                if (cur_high.isGreaterThan(currentExtremePoint)) {
                    currentExtremePoint = cur_high;
                    cur_af = incrementAcceleration(index);
                }
                lastExtreme.put(index, currentExtremePoint);
                lastAf.put(index, cur_af);
            }
        } else { // downtrend
            if (cur_high.isGreaterThanOrEqual(sar)) { // check if switch to up trend
                sar = currentExtremePoint;

                lastAf.put(index, accelerationStart);
                lastExtreme.put(index, cur_high);
                is_up_trend = true;

            } else { // down trend io going on
                if (cur_low.isLessThan(currentExtremePoint)) {
                    currentExtremePoint = cur_low;
                    cur_af = incrementAcceleration(index);
                }
                lastExtreme.put(index, currentExtremePoint);
                lastAf.put(index, cur_af);

            }
        }

        if (is_up_trend) {
            Num lowestPriceOfTwoPreviousBars = new LowestValueIndicator(lowPriceIndicator, 2).getValue(index - 1);
            if (sar.isGreaterThan(lowestPriceOfTwoPreviousBars)) {
                sar = lowestPriceOfTwoPreviousBars;
            }
        } else {
            Num highestPriceOfTwoPreviousBars = new HighestValueIndicator(highPriceIndicator, 2).getValue(index - 1);
            if (sar.isLessThan(highestPriceOfTwoPreviousBars)) {
                sar = highestPriceOfTwoPreviousBars;
            }
        }
        isUpTrendMap.put(index, is_up_trend);
        return sar;
    }
```

Maybe we could introduce some kind of thresholds for such indicators, so that they would only look back for a specific amount of bars that is sufficient to fill the context maps...
